### PR TITLE
fix: onnx PReLU unidirectional broadcasting

### LINKED ISSUE #70570: Incorrect ONNX Export for Unidirectional Broadcasting in PReLU
### 🐛 Describe the bug

```python
import torch
from tvm import relay
import onnx

model = torch.nn.PReLU()

onnx_model = "test.onnx"

torch.onnx.export(model, (torch.tensor(1.0),), onnx_model, verbose=True)

mod, params = relay.frontend.from_onnx(onnx.load(onnx_model))

mod = relay.transform.InferType()(mod)
```

When compiling this **PReLU** operator in ONNX, TVM will fail due to that the given model is incorrect. After some investigation, I found that in ONNX's [specification](https://github.com/onnx/onnx/blob/master/docs/Operators.md#prelu):

> This operator supports unidirectional broadcasting (tensor slope should be **unidirectional broadcastable** to input tensor X);

This means in a valid ONNX model, rank of `slope` must be <= rank of PReLU's input (Unidirectional Broadcasting) while PyTorch's PReLU seems to allow 1D `slope` when input is a scalar. So to make everything compatible, during ONNX exportation, PyTorch end should add an `unsqueeze` operation for the input.

By editing `symbolic_opeset9.py`:

```python
def prelu(g, self, weight):
    self_rank = sym_help._get_tensor_rank(self)
    if self_rank is not None and self_rank > 2:
        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+       return g.op("PRelu", sym_help._unsqueeze_helper(g, self, [0]), weight)
    return g.op("PRelu", self, weight)
```

can fix the issue. :-)

### Versions

PyTorch version: 1.10.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 11.1.0
Clang version: 13.0.0
CMake version: version 3.22.1
Libc version: glibc-2.33

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.10.84-1-MANJARO-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.5.119
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti
Nvidia driver version: 495.44
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.0
/usr/lib/libcudnn_adv_infer.so.8.3.0
/usr/lib/libcudnn_adv_train.so.8.3.0
/usr/lib/libcudnn_cnn_infer.so.8.3.0
/usr/lib/libcudnn_cnn_train.so.8.3.0
/usr/lib/libcudnn_ops_infer.so.8.3.0
/usr/lib/libcudnn_ops_train.so.8.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.0rc3
[pip3] numpydoc==1.1.0
[pip3] torch==1.10.0+cu113
[pip3] torch-tb-profiler==0.3.1
[pip3] torchaudio==0.10.0+cu113
[pip3] torchvision==0.11.1+cu113
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py38h27cfd23_1  
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.1            py38ha9443f7_2  
[conda] mypy_extensions           0.4.3                    py38_0  
[conda] numpy                     1.22.0rc3                pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.10.0+cu113             pypi_0    pypi
[conda] torch-tb-profiler         0.3.1                    pypi_0    pypi
[conda] torchaudio                0.10.0+cu113             pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi

### PR DESCRIPTION
**Pull request by ganler**
Fixes https://github.com/pytorch/pytorch/issues/70570

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
LGTM, could you also add a unit test for this case?

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -719,6 +719,9 @@ def prelu(g, self, weight):
     self_rank = sym_help._get_tensor_rank(self)
     if self_rank is not None and self_rank > 2:
         weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+        self = sym_help._unsqueeze_helper(g, self, [0])
+    assert sym_help._get_tensor_rank(self) >= len(sym_help._get_tensor_sizes(weight)), "slope must be unidirectionally broadcastable to input tensor"
```
Comment (ganler) [inline torch/onnx/symbolic_opset9.py:724]: The bug needs to be reproduced using TVM but will introduce a huge testing dependency. So instead I added a rank checking for it.

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -719,6 +719,9 @@ def prelu(g, self, weight):
     self_rank = sym_help._get_tensor_rank(self)
     if self_rank is not None and self_rank > 2:
         weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+        self = sym_help._unsqueeze_helper(g, self, [0])
+    assert self_rank is None or sym_help._get_tensor_rank(self) >= len(sym_help._get_tensor_sizes(weight)), "slope must be unidirectionally broadcastable to input tensor"
```
Comment (ganler) [inline torch/onnx/symbolic_opset9.py:724]: The bug needs to be reproduced using TVM but will introduce a huge testing dependency. So instead I added a rank checking for it.

**Comment by ganler** [comment]
timeout tests seem to be flaky. pushed an empty commit to trigger the ci.

**Comment by ganler** [comment]
Hi @thiagocrepaldi , could you also double-check the added test case? Thank you!

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -6034,6 +6034,12 @@ def forward(self, x):
                       dynamic_axes={"x": [1, 2]},
                       test_with_inputs=[y])
 
+    def test_prelu_scalar(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6037]: Does this test fail before this change and pass afterwards? If not then I don't think this test is useful.

**Review by garymm — COMMENTED** [review]
I don't think this is actually needed. See my comment here:
https://github.com/pytorch/pytorch/issues/70570#issuecomment-1026258821

**Comment by ganler** [comment]
Hi @garymm , thanks for the review and I have addressed the comments. 

Regarding the benefit of this PR, this change will help:
1. TVM accepts the model in https://github.com/pytorch/pytorch/issues/70570#issue-1092220771
2. ORT reduces the warning of "Expected shape from model of {} does not match actual shape of {1} for output 2"

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
...
-    if self_rank is not None and self_rank > 2:
-        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    if self_rank is not None:
+        if self_rank > 2:
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+        elif self_rank == 0:
+            self = sym_help._unsqueeze_helper(g, self, [0])
+            self_rank = sym_help._get_tensor_rank(self)
+
+    slope_rank = sym_help._get_tensor_rank(weight)
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:727]: Don't mix torch names and ONNX names, it's confusing.

```suggestion
    weight_rank = sym_help._get_tensor_rank(weight)
```

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
...
 def prelu(g, self, weight):
     self_rank = sym_help._get_tensor_rank(self)
-    if self_rank is not None and self_rank > 2:
-        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    if self_rank is not None:
+        if self_rank > 2:
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+        elif self_rank == 0:
+            self = sym_help._unsqueeze_helper(g, self, [0])
+            self_rank = sym_help._get_tensor_rank(self)
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:725]: Don't we know this for sure?
```suggestion
            self_rank = 1
```

File: torch/onnx/symbolic_opset9.py [hunk 6]
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
 
 def prelu(g, self, weight):
     self_rank = sym_help._get_tensor_rank(self)
-    if self_rank is not None and self_rank > 2:
-        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    if self_rank is not None:
+        if self_rank > 2:
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:722]:
```suggestion
            # make weight unidirectional broadcastable
            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
```

File: torch/onnx/symbolic_opset9.py [hunk 7]
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
...
+        elif self_rank == 0:
+            self = sym_help._unsqueeze_helper(g, self, [0])
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:724]:
```suggestion
            # weight is always rank 1. torch allows scalar self, and ONNX is ambiguous
            # about whether this is allowed, but some implementations enforce
            # rank(self) >= rank(weight), which makes sense.
            self = sym_help._unsqueeze_helper(g, self, [0])
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -6034,6 +6034,7 @@ def forward(self, x):
                       dynamic_axes={"x": [1, 2]},
                       test_with_inputs=[y])
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6035]: Add a test that uses scalar inputs to trigger the new code path.
Even if the difference between the old behavior and new behavior is just "ORT doesn't print a warning", that's OK. Better to have a test that executes the code path and makes sure it basically works.

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks for the contribution, and sorry for the latency.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 9]
```diff
@@ -6034,6 +6034,11 @@ def forward(self, x):
...
 
+    def test_prelu_scalar(self):
+        x = torch.scalar_tensor(1.)
+        self.run_test(torch.nn.PReLU(), x, input_names=["x"],
+                      test_with_inputs=[x])
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6040]: Remove test_with_inputs. By default `run_test` will test with the inputs in the second argument, so this is redundant.

**Review by garymm — CHANGES_REQUESTED** [review]

**Review by garymm — APPROVED** [review]
Will merge once checks pass.

**Comment by garymm** [comment]
@msaroufim could you please import? (non-MS contribution, can't use merge bot yet)

**Pull request closed without merging**